NCT38xx multi-function device

keith-zephyr opened the pull request:
The Nuvoton NCT38xx provides both a TCPC (Type C port controller) and I/O expander capabilities.

Add a multi-function device (MFD) driver parent to the existing nct38xx GPIO devices to coordinate I2C register accesses from multiple driver types.  Support for the TCPC portion of the NCT38xx will be added later.

keith-zephyr committed:
mfd: Add NCT38xx multi-function driver

The Nuvoton NCT38xx is a multi-function device providing a TCPC
controller and a I/O expander (GPIO driver).  Add a multi-function
driver to manage exclusive access to the device.

Tested with "twister -T tests/drivers/build_all/gpio".

Signed-off-by: Keith Short <keithshort@google.com>

keith-zephyr committed:
gpio: nct38xx: Cleanup semaphore usage

Cleanup the semaphore usage in the NCT38xx GPIO driver.

Signed-off-by: Keith Short <keithshort@google.com>

ChiHuaL commented on drivers/mfd/mfd_nct38xx.c:
```diff
@@ -0,0 +1,138 @@
+#define MFD_NCT38XX_DEFINE(inst)                                                                   \
+	static struct mfd_nct38xx_data nct38xx_data_##inst;                                        \
+	static const struct mfd_nct38xx_config nct38xx_cfg_##inst = {                              \
+		.i2c_dev = I2C_DT_SPEC_INST_GET(inst),                                             \
```
Should we check the readiness of .i2c_dev?

ChiHuaL commented on drivers/gpio/Kconfig.nct38xx:
```diff
@@ -6,26 +6,27 @@
 	default y
-	depends on DT_HAS_NUVOTON_NCT38XX_GPIO_PORT_ENABLED
-	depends on I2C
+depends on DT_HAS_NUVOTON_NCT38XX_GPIO_PORT_ENABLED
```
Miss TAB at  the beginning of this line.

fabiobaltieri commented on drivers/mfd/mfd_nct38xx.c:
```diff
@@ -0,0 +1,138 @@
+int nct38xx_reg_read_byte(const struct device *dev, uint8_t reg_addr, uint8_t *val)
+{
+	const struct mfd_nct38xx_config *const config =
+		(const struct mfd_nct38xx_config *)dev->config;
```
You can drop these castings, saves a line wrap, they don't add anything anyway.

gmarull commented on drivers/mfd/mfd_nct38xx.c:
```diff
@@ -0,0 +1,138 @@
+	return 0;
+}
+
+int mfd_nct38xx_lock(const struct device *dev, k_timeout_t timeout)
+{
+	struct mfd_nct38xx_data *data = dev->data;
+
+	return k_sem_take(&data->lock, timeout);
+}
+
+void mfd_nct38xx_unlock(const struct device *dev)
+{
+	struct mfd_nct38xx_data *data = dev->data;
+
+	k_sem_give(&data->lock);
+}
```
this seems like just a wrapper around  `k_sem_take/give`. Wouldn't it be better to allow getting a reference to a shared lock..? So drivers don't need to use a new idiom for locking.

keith-zephyr replied:
Changed to return a reference the semaphore.

gmarull commented on drivers/mfd/mfd_nct38xx.c:
```diff
@@ -0,0 +1,138 @@
+ *
+ * @return 0 if successful, otherwise failed.
+ */
+int nct38xx_reg_read_byte(const struct device *dev, uint8_t reg_addr, uint8_t *val)
+{
+	const struct mfd_nct38xx_config *const config =
+		(const struct mfd_nct38xx_config *)dev->config;
+	return i2c_reg_read_byte_dt(&config->i2c_dev, reg_addr, val);
+}
```
what is the point of a plain wrapper around i2c_reg_read?

gmarull replied:
but again, what is the point of plain wrappers around i2c? just take i2c parent instance on each subdevice and use the standard i2c API (see e.g. npm6001). It would only make sense if i2c ops require extra steps/crafting like in npm1300, to guarantee atomic operations, etc.

fabiobaltieri replied:
They are probably a leftover from before we had `_dt` macros, may make sense to do a swipe of i2c drivers and drop them where applicable, sure there's few others.

keith-zephyr replied:
I removed the wrapper routines, but kept the changes in a separate commit.  There are some non-trivial conversions to switch from `nct38xx_reg_update()` to `i2c_reg_update_byte_dt()`.

gmarull commented on drivers/mfd/mfd_nct38xx.c:
```diff
@@ -0,0 +1,138 @@
+		(const struct mfd_nct38xx_config *)dev->config;
+	return i2c_reg_write_byte_dt(&config->i2c_dev, reg_addr, val);
+}
+
+/**
+ * @brief Write a sequence of NCT38XX registers
+ *
+ * @param dev NCT38XX device
+ * @param start_addr The register start address
+ * @param buf A pointer to a buffer for the data to write
+ * @param num_bytes Number of data to write
+ *
+ * @return 0 if successful, otherwise failed.
+ */
+int nct38xx_reg_burst_write(const struct device *dev, uint8_t start_addr, uint8_t *buf,
+			    uint32_t num_bytes)
+{
+	const struct mfd_nct38xx_config *const config =
+		(const struct mfd_nct38xx_config *)dev->config;
+	return i2c_burst_write_dt(&config->i2c_dev, start_addr, buf, num_bytes);
+}
+
+/**
+ * @brief Compare data & write a NCT38XX register
+ *
+ * @param dev NCT38XX device
+ * @param reg_addr Register address
+ * @param reg_val Old register data
+ * @param new_val New register data
+ *
+ * @return 0 if successful, otherwise failed.
+ */
+int nct38xx_reg_update(const struct device *dev, uint8_t reg_addr, uint8_t reg_val, uint8_t new_val)
+{
+	if (reg_val == new_val) {
+		return 0;
+	}
+
+	return nct38xx_reg_write_byte(dev, reg_addr, new_val);
+}
```
same here. Also, docs are put in the wrong place (.c)

keith-zephyr commented:
Commit https://github.com/zephyrproject-rtos/zephyr/commit/1c95a6ad59e311a73c241af2f28dca209b380031 is a rebase only.

keith-zephyr committed:
gpio: nct38xx: Remove reg access wrappers

Remove the register access wrappers used to read/write the NCT38xx
register space.

Signed-off-by: Keith Short <keithshort@google.com>

gmarull commented on include/zephyr/drivers/mfd/nct38xx.h:
```diff
@@ -0,0 +1,38 @@
+#ifndef ZEPHYR_INCLUDE_DRIVERS_MFD_NCT38XX_H_
+#define ZEPHYR_INCLUDE_DRIVERS_MFD_NCT38XX_H_
+
+#include <zephyr/sys_clock.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief Get the semaphore reference for a NCT38xx instance. Callers
+ * should pass the return value to k_sem_take/k_sem_give
+ *
+ * @param[in] dev     Pointer to device struct of the driver instance
+ *
+ * @return Address of the semaphore
+ */
+struct k_sem *mfd_nct38xx_get_lock_reference(const struct device *dev);
+
+/**
+ * @brief Get the I2C DT spec reference for a NCT38xx instance.
+ *
+ * @param[in] dev     Pointer to device struct of the driver instance
+ *
+ * @return Address of the I2C DT spec
+ */
+const struct i2c_dt_spec *mfd_nct38xx_get_i2c_dt_spec(const struct device *dev);
```
this header is not self-contained, missing: kernel.h, i2c.h, device.h

keith-zephyr replied:
Fixed.

keith-zephyr commented:
https://github.com/zephyrproject-rtos/zephyr/commit/b725e79f9ed7f80f648428120b79e2b0d8a73e62 is a rebase only.
